### PR TITLE
check the executor subject permission for action apply

### PR DESCRIPTION
**Pull request by zhujian7**
issue reference: https://github.com/open-cluster-management-io/work/issues/153

### CODE REVIEW

File: pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go [hunk 1]
```diff
@@ -154,6 +158,9 @@ func (m *ManifestWorkController) sync(ctx context.Context, controllerContext fac
 	errs := []error{}
 	// Apply resources on spoke cluster.
 	resourceResults := make([]applyResult, len(manifestWork.Spec.Workload.Manifests))
+	// check the Executor subject permission before applying
+	resourceResults = m.validateManifests(ctx, manifestWork.Spec.Executor,
```
Comment (qiujian16) [inline pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go]: shouldn't we validate each manifest separately?

File: pkg/spoke/helper/helper.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+// DeleteAppliedResources deletes all given applied resources and returns those pending for finalization
+// If the uid recorded in resources is different from what we get by client, ignore the deletion.
+func DeleteAppliedResources(
+	ctx context.Context,
+	resources []workapiv1.AppliedManifestResourceMeta,
+	reason string,
+	dynamicClient dynamic.Interface,
+	recorder events.Recorder,
+	owner metav1.OwnerReference,
+	kubeClient kubernetes.Interface, executor *workapiv1.ManifestWorkExecutor) ([]workapiv1.AppliedManifestResourceMeta, []error) {
```
Comment (qiujian16) [inline pkg/spoke/helper/helper.go]: I think you want to input validator here, we need to consider cache in the future.

File: pkg/spoke/helper/helper.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+		err = validator.ValidateDelete(ctx, executor, gvr, resource.Namespace, resource.Name)
+		if err != nil {
+			if !apierrors.IsForbidden(err) {
+				errs = append(errs, fmt.Errorf(
+					"failed to check delete permission for resource %v with key %s/%s: %w",
+					gvr, resource.Namespace, resource.Name, err))
+				continue
+			}
+			// TODO: consider to reflect the not allowed error on the manifest work condition
```
Comment (qiujian16) [inline pkg/spoke/helper/helper.go]: this will be certainly needed. I actually think all delete action should be put in the separate PR together with how we handle condition.

File: pkg/spoke/controllers/finalizercontroller/appliedmanifestwork_finalize_controller.go [hunk 4]
```diff
@@ -96,12 +105,25 @@ func (m *AppliedManifestWorkFinalizeController) syncAppliedManifestWork(ctx cont
...
+		return err
+	}
+
 	// Work is deleting, we remove its related resources on spoke cluster
 	// We still need to run delete for every resource even with ownerref on it, since ownerref does not handle cluster
 	// scoped resource correctly.
 	reason := fmt.Sprintf("manifestwork %s is terminating", appliedManifestWork.Spec.ManifestWorkName)
-	resourcesPendingFinalization, errs := helper.DeleteAppliedResources(
-		ctx, appliedManifestWork.Status.AppliedResources, reason, m.spokeDynamicClient, controllerContext.Recorder(), *owner)
+	resourcesPendingFinalization, errs := spokehelper.DeleteAppliedResources(
```
Comment (qiujian16) [inline pkg/spoke/controllers/finalizercontroller/appliedmanifestwork_finalize_controller.go]: we also need to update condition for work here, to declare the remaining resource and err

File: pkg/spoke/controllers/finalizercontroller/manifestwork_finalize_controller.go [hunk 5]
```diff
@@ -68,14 +68,17 @@ func (m *ManifestWorkFinalizeController) sync(ctx context.Context, controllerCon
...
 		if err != nil {
 			return err
 		}
 	case err != nil:
 		return err
 	case !manifestWork.DeletionTimestamp.IsZero():
-		err := m.deleteAppliedManifestWork(ctx, appliedManifestWorkName)
+		// delete the applied manifestwork, but orphan the dependent resources, we will
+		// check the executor subject permission before deleting these resources
+		orphanDependents := manifestWork.Spec.Executor != nil
```
Comment (qiujian16) [inline pkg/spoke/controllers/finalizercontroller/manifestwork_finalize_controller.go]: hrm If we orphan delete the appliedwork, will lose the ownership of the resources, right? it could be conflict with deleteOption.

**Comment by zhujian7** [comment]
@qiujian16 @elgnay PTAL

File: pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go [hunk 6]
```diff
@@ -187,18 +188,41 @@ func (m *ManifestWorkController) sync(ctx context.Context, controllerContext fac
 		manifestCondition.Conditions = append(manifestCondition.Conditions, buildAppliedStatusCondition(result))
 
 		newManifestConditions = append(newManifestConditions, manifestCondition)
+
+		if apierrors.IsForbidden(result.Error) {
+			// If it is a forbidden error, after the condition is constructed, we set the error to nil
```
Comment (qiujian16) [inline pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go]: I think we should let validator to return the requeue time in the error. We can construct a special error struct in validator. Here we get the requeue time for each manifest in the work, and find the min one. This will be usefule when we implement validate cache later. Because the validator should use cache ttl to calculate the requeue time.

File: pkg/spoke/auth/auth.go [hunk 7]
```diff
@@ -33,6 +33,15 @@ type ExecutorValidator interface {
 		gvr schema.GroupVersionResource, namespace, name string, action ExecuteAction) error
 }
 
+type NotAllowedError struct {
+	Err         error
+	RequeueTime time.Duration
+}
+
+func (e *NotAllowedError) Error() string {
+	return e.Err.Error()
```
Comment (qiujian16) [inline pkg/spoke/auth/auth.go]: nit: would you also add requeue time in the error message?

File: pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go [hunk 8]
```diff
@@ -189,12 +189,16 @@ func (m *ManifestWorkController) sync(ctx context.Context, controllerContext fac
 
 		newManifestConditions = append(newManifestConditions, manifestCondition)
 
-		if apierrors.IsForbidden(result.Error) {
-			// If it is a forbidden error, after the condition is constructed, we set the error to nil
-			// and requeue the item after a little while
-			needRequeue = true
+		// If it is a forbidden error, after the condition is constructed, we set the error to nil
+		// and requeue the item
+		var autoError = &auth.NotAllowedError{}
```
Comment (qiujian16) [inline pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go]: auto or notAllowedError?

**Review by qiujian16 — COMMENTED** [review]
just two nit comment, others looks good

/approve

File: pkg/spoke/auth/auth.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	sa := executor.Subject.ServiceAccount
+	if sa == nil {
+		return fmt.Errorf("the executor service account is nil")
+	}
+
+	var verbs []string
+	switch action {
+	case ApplyAction:
```
Comment (qiujian16) [inline pkg/spoke/auth/auth.go:76]: list is not necessary, right?

File: deploy/spoke/clusterrole.yaml [hunk 10]
```diff
@@ -15,4 +15,6 @@ rules:
 - apiGroups: ["work.open-cluster-management.io"]
   resources: ["appliedmanifestworks/finalizers"]
   verbs: ["update"]
-  
\ No newline at end of file
+- apiGroups: ["authorization.k8s.io"]
```
Comment (qiujian16) [inline deploy/spoke/clusterrole.yaml:19]: nit: you may want to add this in operator at first.

File: pkg/spoke/auth/auth.go (hunk 9 above)
Comment (zhujian7) [reply pkg/spoke/auth/auth.go:76]: I think it is not necessary, but the [proposal](https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/34-work-executor-group#workflow) has the "list" verb, so do we need to update the proposal?

File: deploy/spoke/clusterrole.yaml (hunk 10 above)
Comment (zhujian7) [reply deploy/spoke/clusterrole.yaml:19]: yes, I  added it https://github.com/open-cluster-management-io/registration-operator/pull/273

File: pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go (hunk 8 above)
Comment (zhujian7) [reply pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go]: sorry, typo, changed to authError

File: pkg/spoke/auth/auth.go (hunk 7 above)
Comment (zhujian7) [reply pkg/spoke/auth/auth.go]: added

File: pkg/spoke/auth/auth.go (hunk 9 above)
Comment (qiujian16) [reply pkg/spoke/auth/auth.go:76]: yes, I think so.
